### PR TITLE
Normalize address and Infer protocol

### LINKED ISSUE #185: Infer the "Detail of a loan" parameters
Steps:

0) Read our [Contributor Guidelines](https://github.com/CarmineOptions/derisk-research/blob/master/CONTRIBUTING.md) and [README](https://github.com/CarmineOptions/derisk-research/blob/master/README.md).
1) Check out the current implementation of our streamlit dashboard, specifically the part that displays the details of a specific loan: [https://github.com/CarmineOptions/derisk-research/blob/master/app.py#L393](https://github.com/CarmineOptions/derisk-research/blob/master/app.py#L393). The dashboard can be found [here](https://starknet.derisk.carmine.finance/).
2) Under "Detail of a loan", anyone can search for statistics of a loan specified by the address of the user and the protocol name. However, if the combination of the address and protocol doesn't exist, i.e., there is no loan for the given user under the given protocol, because
a) the user's address is correct, but missing the leading zeroes, or
image
b) the protocol name is correct, but not exactly matching either "zkLend", "Nostra Alpha", or "Nostra Mainnet"
image
the dashboard crashes on an "IndexError" error, and the loan is not found or displayed. Run the dashboard locally and try it yourself.
3) Handle such cases by 
a) adding the leading zeroes to the user's address, or
b) infering the correct protocol name
to match the values in the "loans_data" dataframe, so that the loan details are displayed in these scenarios.

### PR DESCRIPTION
**Pull request by od-hunter**
Resolves  #185

@djeck1432 , Kindly review. Changes made to app.py, to normalize user address and infer protocol.

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -16,7 +16,21 @@
 import src.settings
 import src.swap_amm
 
+import difflib
 
+# Function to normalize user address by adding leading zeroes if necessary
+def normalize_address(address, expected_length=42):
```
Comment (djeck1432) [inline app.py:22]: please, add annotation types and return type
and docstring

File: app.py [hunk 2]
```diff
@@ -16,7 +16,21 @@
...
+    if len(address) < expected_length:
+        address = address.zfill(expected_length)
+    return address
+
+# Function to find the closest protocol match using fuzzy matching
+def infer_protocol_name(input_protocol, valid_protocols):
```
Comment (djeck1432) [inline app.py:28]: please, add annotation types and return type
and docstring

File: app.py [hunk 3]
```diff
@@ -16,7 +16,21 @@
...
+    closest_match = difflib.get_close_matches(input_protocol, valid_protocols, n=1, cutoff=0.6)
+    if closest_match:
```
Comment (djeck1432) [inline app.py:30]: Instead of 4 lines of code you can put it in one line:
```
return closest_match and closest_match[0] or input_protocol
```

**Review by djeck1432 — CHANGES_REQUESTED** [review]

**Review by djeck1432 — APPROVED** [review]
Good job

**Comment by od-hunter** [comment]
> Good job

Thank you ser🫡

File: app.py [hunk 4]
```diff
@@ -408,6 +437,13 @@ def _get_risk_level(debt_to_supply_ratio: float) -> str:
         if not protocol:
             streamlit.write(f'Selected random protocol = {random_protocol}.')
             protocol = random_protocol
+        # Normalize the user address by adding leading zeroes if necessary
+        user = normalize_address(user)
```
Comment (lukaspetrasek) [inline app.py:441]: Pls use the already existing function: https://github.com/CarmineOptions/derisk-research/blob/master/src/helpers.py#L175.

File: app.py [hunk 5]
```diff
@@ -16,7 +16,36 @@
 import src.settings
 import src.swap_amm
 
+import difflib
```
Comment (lukaspetrasek) [inline app.py:19]: Move the import to standard library imports, mind the alphabetical sorting.

**Comment by lukaspetrasek** [comment]
Hi @od-hunter , added 2 comments 🙏🏼

**Comment by od-hunter** [comment]
> Hi @od-hunter , added 2 comments 🙏🏼

@lukaspetrasek , all done ser!

**Comment by djeck1432** [comment]
@od-hunter Could you please run `black` and `isort` as it's described here https://github.com/CarmineOptions/derisk-research/blob/master/PR_RULES.md
I can see in your PR some formatting issue, thank you

**Comment by od-hunter** [comment]
> @od-hunter Could you please run `black` and `isort` as it's described here https://github.com/CarmineOptions/derisk-research/blob/master/PR_RULES.md I can see in your PR some formatting issue, thank you

Hello ser @djeck1432 , please is it app.py that you want me to format? (since that is the file I made changes to)

**Comment by lukaspetrasek** [comment]
Hi @od-hunter , could you pls resolve the conflicts? 🙏🏼

File: app.py [hunk 6]
```diff
@@ -437,8 +421,9 @@ def _get_risk_level(debt_to_supply_ratio: float) -> str:
         if not protocol:
             streamlit.write(f'Selected random protocol = {random_protocol}.')
             protocol = random_protocol
+            
         # Normalize the user address by adding leading zeroes if necessary
-        user = normalize_address(user)
+        user =src.helpers.add_leading_zeros(user)
```
Comment (lukaspetrasek) [inline app.py:426]:
```suggestion
        user = src.helpers.add_leading_zeros(user)
```

File: app.py [hunk 7]
```diff
@@ -16,23 +17,6 @@
 import src.settings
 import src.swap_amm
```
Comment (lukaspetrasek) [inline app.py:19]: Add an empty line below.

**Comment by od-hunter** [comment]
@lukaspetrasek  conflict resolved

File: app.py [hunk 8]
```diff
@@ -311,60 +389,56 @@ def main():
...
             f"lending protocols is {_get_risk_level(example_row['debt_to_supply_ratio'])}."
-        )    
+        )
         streamlit.write(
             f"The ratio of liquidated debt to available supply is {round(example_row['debt_to_supply_ratio'] * 100)}%.Debt"
             f" worth of {int(example_row['liquidable_debt_at_interval']):,} USD will be liquidated while the AMM swaps "
             f"capacity will be {int(example_row['debt_token_supply']):,} USD."
         )
 
-    streamlit.header("Liquidable debt")
```
Comment (lukaspetrasek) [inline app.py:357]: Seems like you've accidentally deleted valid code!

File: app.py [hunk 9]
```diff
@@ -406,82 +484,89 @@ def _get_risk_level(debt_to_supply_ratio: float) -> str:
...
             protocol = random_protocol
 
-    loan = loans_data.loc[
-        (loans_data['User'] == user)
-        & (loans_data['Protocol'] == protocol),
-    ]
+        # Normalize the user address by adding leading zeroes if necessary
+        user = src.helpers.add_leading_zeros(user)
 
-    if loan.empty:
```
Comment (lukaspetrasek) [inline app.py:431]: DITTO

File: app.py (hunk 8 above)
Comment (od-hunter) [reply app.py:357]: "@lukaspetrasek, I am currently facing some challenges with resolving the conflicts in this PR. Would it be acceptable to close this PR, re-implement my changes, and submit a new one?"

**Comment by od-hunter** [comment]
@djeck1432 , @lukaspetrasek please review

**Pull request merged**